src/lem/mod.rs

```diff
@@ -167,7 +167,7 @@ impl LEMOP {
     /// Note: this function is not supposed to be called manually. It's used by
     /// `LEM::new`, which is the API that should be used directly.
     pub fn deconflict(
-        &self,
+        self,
```

Comment:
Is it causing the entire `LEM` to be copied?

Reply:
No, but it does something with the owned value you already require (without consuming it, before this PR) in `new`

Reply:
Your LEMOP type is not Copy, so copies of that type will be explicit and denotes by `foo.clone()`. 

The only place your deconflict function is called is the `lem::new` function, which takes a LEMOP by value without much justification for it, because the only place that actually refers to the LEMOP is here, and it *copies* all it needs from a reference to the LEMOP.  That's explicit for the name, and implicit for the Tag (which is Copy).

I would suggest looking at C-CALLER-CONTROL in the rust API guidelines.

Reply:
> The only place your deconflict function is called is the `lem::new` function

What about the recursive calls?

Reply:
> If a function does not require ownership of an argument, it should take a shared or exclusive borrow of the argument rather than taking ownership and dropping the argument.

`deconflict` doesn't need ownership of the `LEMOP`. It's never cloned

Reply:
In the interest of expediency, I’ll remove the last few commits of this PR, thereby not changing the way `deconflict` functions. I maintain a few claims:
1. the call to deconflict as modified by this PR does not create clones of the `LEMOP`: `LEMOP` is not `Copy`, therefore it’s impossible to clone it wholesale without calling `.clone()`, and I have introduced no such `.clone` call in the present PR,
2 `deconflict` does copy significant portions of the content of a `LEMOP`, e.g. the tags (which are `Copy`), and other components as well (e.g. `ptr.name().clone()`)
3. `LEM::new`, before this PR, takes ownership of its argument `lem_op` which it does not need, hinting that the elements of that `lem_op` that will end up in the created `lem` are obtained through copy (see 1.)

Reply:
Thanks, I will fix `LEM::new` so that it doesn't take ownership of the argument `lem_op`

Reply:
Okay, I'm confused now. `LEM::new` is supposed to create a `LEMOP` based on a `LEMOP` that's a literal reference: we look at it and then we create a modified version of it.

I think I'm mixing up the notion of "read-only" with "borrowing"

Reply:
This is the piece of writing that should provide a better mental model: https://docs.rs/dtolnay/0.0.11/dtolnay/macro._02__reference_types.html

It may well be that the answer is indeed what you suggest: `LEM::new` (and resp. `LEMOP::deconflict`) takes its argument (resp receiver) by reference, and that's it. The question is more to think about how there's memory allocated somewhere for the purpose of representing the data contained in the LEMOP. Could deconflict make a good use of it (e.g. because it's going to put those allocated bits in a HashMap)? If so, is it likely that no one else needs a reference to those bits?

If the answer to both questions is yes (and I'm not sure it is), it's probably a sign that one or both of those functions should take ownership of their argument, thereby doing less allocation themselves, and letting the caller figure out how they want to handle that.

The later is important because, well, there are many ways to be clever about allocation.